common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -0,0 +1,959 @@
+      this.encounteredFailure = false;
+      this.mapTracker = new RoaringBitmap();
+      this.chunkTracker = new RoaringBitmap();
+    }
+
+    public long getPosition() {
+      return position;
+    }
+
+    public void setPosition(long position) {
+      logger.trace("{} shuffleId {} reduceId {} current pos {} update pos {}", appShuffleId.appId,
+        appShuffleId.shuffleId, reduceId, this.position, position);
+      this.position = position;
+    }
+
+    boolean isEncounteredFailure() {
+      return encounteredFailure;
+    }
+
+    void setEncounteredFailure(boolean encounteredFailure) {
+      this.encounteredFailure = encounteredFailure;
+    }
+
+    int getCurrentMapIndex() {
+      return currentMapIndex;
+    }
+
+    void setCurrentMapIndex(int mapIndex) {
+      logger.trace("{} shuffleId {} reduceId {} updated mapIndex {} current mapIndex {}",
+        appShuffleId.appId, appShuffleId.shuffleId, reduceId, currentMapIndex, mapIndex);
+      this.currentMapIndex = mapIndex;
+    }
+
+    long getLastChunkOffset() {
+      return lastChunkOffset;
+    }
+
+    void blockMerged(int mapIndex) {
+      logger.debug("{} shuffleId {} reduceId {} updated merging mapIndex {}", appShuffleId.appId,
+        appShuffleId.shuffleId, reduceId, mapIndex);
+      mapTracker.add(mapIndex);
+      chunkTracker.add(mapIndex);
+      lastMergedMapIndex = mapIndex;
+    }
+
+    void resetChunkTracker() {
+      chunkTracker.clear();
+    }
+
+    /**
+     * Appends the chunk offset to the index file and adds the map index to the chunk tracker.
+     *
+     * @param chunkOffset the offset of the chunk in the data file.
+     * @param mapIndex the map index to be added to chunk tracker.
+     */
+    void updateChunkInfo(long chunkOffset, int mapIndex) throws IOException {
+      long idxStartPos = -1;
+      try {
+        // update the chunk tracker to meta file before index file
+        writeChunkTracker(mapIndex);
+        idxStartPos = indexChannel.position();
+        logger.trace("{} shuffleId {} reduceId {} updated index current {} updated {}",
+          appShuffleId.appId, appShuffleId.shuffleId, reduceId, this.lastChunkOffset,
+          chunkOffset);
+        indexWriteStream.writeLong(chunkOffset);
+      } catch (IOException ioe) {
+        if (idxStartPos != -1) {
+          // reset the position to avoid corrupting index files during exception.
+          logger.warn("{} shuffleId {} reduceId {} reset index to position {}",
+            appShuffleId.appId, appShuffleId.shuffleId, reduceId, idxStartPos);
+          indexChannel.position(idxStartPos);
+        }
+        throw ioe;
+      }
+      this.lastChunkOffset = chunkOffset;
+    }
+
+    private void writeChunkTracker(int mapIndex) throws IOException {
+      if (mapIndex == -1) {
+        return;
+      }
+      chunkTracker.add(mapIndex);
+      if (trackerBuf == null) {
+        trackerBuf = Unpooled.buffer(Encoders.Bitmaps.encodedLength(chunkTracker));
+      }
```

Comment:
```suggestion
      trackerBuf.ensureWritable(Encoders.Bitmaps.encodedLength(chunkTracker))；
```

hmm...shall we call `ensureWritable` here? I think the `buf` inside `encode()` is always supposed to be capable. We'd better not break the assumption. @otterc

Reply:
That will calculate the length of the bitmap twice every time the chunk is serialized. Also, I don't think that there is an assumption that the `buf` cannot be expanded. If the type of buf allows expansion it would. If you look at the encode of ByteArrays, even that uses `    buf.writeBytes(...)` which will call `ensureWritable(length)`. The difference here is that we changed it to use `RoaringBitmap.serialize` instead of `AbstractByteBuf.writeBytes`.

Reply:
> That will calculate the length of the bitmap twice every time the chunk is serialized. 

Sorry, which two?

> If you look at the encode of ByteArrays, even that uses buf.writeBytes(...) which will call ensureWritable(length). 

I think it's different. First, the invocation of `ensureWritable()` of `ByteArrays ` is not decided by Spark itself but Netty.
And I'd expect the invocation of `ensureWritable()` here is actually NoOp because I believe we always allocate the exact buffer according to the `encodedLength`. Second, you can't find `ensureWritable` in other places in `Encoders`. That's also why I believe we always guarantee the exact buffer.

cc @mridulm @tgravescs @attilapiros for more inputs on this problem. Related context: https://github.com/apache/spark/pull/30062#issuecomment-722502275

Reply:
> Sorry, which two?

1.  `trackerBuf.ensureWritable(Encoders.Bitmaps.encodedLength(chunkTracker)` will call `b.serializedSizeInBytes()`
2. `Encoders.Bitmaps.encode` will call `b.serializedSizeInBytes()` again.

I am not sure how expensive `RoaringBitmap.serializedSizeInBytes` is, but my preference is to avoid computing that twice.

> I think it's different. First, the invocation of ensureWritable() of ByteArrays is not decided by Spark itself but Netty.

But the API of Spark Encoders  uses `io.netty.buffer.ByteBuf` and doesn't restrict it to a non-expandable ByteBuf implementation.  
For all the other types we rely on netty's method of writing. This is the only place where we used `RoaringBitmap.serialize`. This is different from the rest of encoders. If adding `ensureWritable` here helps with not calculating the length multiple times, then why not?

Reply:
What about not using the `Encoders` when the bitmap is written to/read from a file but `DataOutputStream` would be used instead, like: 
https://github.com/RoaringBitmap/RoaringBitmap/blob/a3764b0336fcdd351cc61fab00ef5ab125c091a5/examples/src/main/java/SerializeToDiskExample.java#L26-L29

and 

https://github.com/RoaringBitmap/RoaringBitmap/blob/a3764b0336fcdd351cc61fab00ef5ab125c091a5/examples/src/main/java/SerializeToDiskExample.java#L36-L39



This way `ensureWritable` is not needed as when it is encoded for a message (sent via Netty) as the buffer size is accurately calculated.

Reply:
I think @attilapiros's suggestion makes sense.
We can just add a file based serialization API inside `Encoders.Bitmaps` to consolidate all related logic at one place.
We do not need a file based deserialization API though.
The metaFile is transferred from server side to the client side leveraging Netty zero copy transfer, thus the server side does not need to perform any deserialization.
The client side will receive the data already as Netty ByteBuf, and can use existing `Bitmaps#decode` for deserialization.
This code is already committed in #29855 
https://github.com/apache/spark/blob/master/common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/MergedBlockMeta.java#L55

By doing this, we can get rid of `trackerBuf` and all the associated issues discussed here.

Reply:
I will be incorporating this suggestion when writing the `chunkTracker` to file. This also avoids using an additional bytebuf when writing to the file. 
I also discussed with @mridulm and @Victsm that whenever we encode a RoaringBitmap in a message `RoaringBitmap.serializedSizeInBytes` is called twice per bitmap. Based on @mridulm suggestion, will update `Encoders.Bitmap.encode`  to 
```
    public static void encode(ByteBuf buf, RoaringBitmap b) {
      int length = buf.writableBytes();
      // RoaringBitmap requires nio ByteBuffer for serde. We expose the netty ByteBuf as a nio
      // ByteBuffer. Here, we need to explicitly manage the index so we can write into the
      // ByteBuffer, and the write is reflected in the underneath ByteBuf.
      ByteBuffer byteBuffer = buf.nioBuffer(buf.writerIndex(), length);
      b.serialize(byteBuffer);
      buf.writerIndex(buf.writerIndex() + byteBuffer.position());
    }
```

Reply:
@Ngone51 @attilapiros @Victsm @mridulm I have updated the PR with these changes. Please take a look when you get a chance.

Reply:
thanks, looks good.